pubsub/natspubsub/nats.go

```diff
@@ -0,0 +1,292 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package natspubsub provides a pubsub implementation for NATS.io.
+// Use CreateTopic to construct a *pubsub.Topic, and/or CreateSubscription
+// to construct a *pubsub.Subscription. This package uses msgPack and the
+// ugorji driver to encode and decode driver.Message to []byte.
+//
+// As
+//
+// natspubsub exposes the following types for As:
+//  - Topic: *nats.Conn
+//  - Subscription: *nats.Subscription
+//  - Message: *nats.Msg
+
+package natspubsub // import "gocloud.dev/pubsub/natspubsub"
+
+import (
+	"context"
+	"errors"
+	"reflect"
+
+	"github.com/nats-io/go-nats"
+	"github.com/ugorji/go/codec"
+	"gocloud.dev/gcerrors"
+	"gocloud.dev/pubsub"
+	"gocloud.dev/pubsub/driver"
+)
+
+var errNotInitialized = errors.New("natspubsub: topic not initialized")
+
+type topic struct {
+	nc   *nats.Conn
+	subj string
+}
+
+// For encoding we use msgpack from github.com/ugorji/go.
+// It was already imported in go-cloud and is reasonably performant.
+// However the more recent version has better resource handling with
+// the addition of the following:
+// 	mh.ExplicitRelease = true
+// 	defer enc.Release()
+// However this is not compatible with etcd at the moment.
+// https://github.com/etcd-io/etcd/pull/10337
+var mh codec.MsgpackHandle
+
+func init() {
+	// driver.Message.Metadata type
+	dm := driver.Message{}
+	mh.MapType = reflect.TypeOf(dm.Metadata)
+}
+
+// We define our own version of message here for encoding that
+// only encodes Body and Metadata. Otherwise we would have to
+// add codec decorations to driver.Message.
+type encMsg struct {
+	Body     []byte            `codec:",omitempty"`
+	Metadata map[string]string `codec:",omitempty"`
+}
+
+// CreateTopic returns a *pubsub.Topic for use with NATS.
+// We delay checking for the proper syntax here.
+func CreateTopic(nc *nats.Conn, topicName string) *pubsub.Topic {
+	return pubsub.NewTopic(createTopic(nc, topicName))
+}
+
+// createTopic returns the driver for CreateTopic. This function exists so the test
+// harness can get the driver interface implementation if it needs to.
+func createTopic(nc *nats.Conn, topicName string) driver.Topic {
+	return &topic{nc, topicName}
+}
+
+// SendBatch implements driver.Topic.SendBatch.
+func (t *topic) SendBatch(ctx context.Context, msgs []*driver.Message) error {
+	if t == nil || t.nc == nil {
+		return errNotInitialized
+	}
+
+	// Reuse if possible.
+	var em encMsg
+	var raw [1024]byte
+	b := raw[:0]
+	enc := codec.NewEncoderBytes(&b, &mh)
+
+	for _, m := range msgs {
+		if err := ctx.Err(); err != nil {
+			return err
+		}
+		enc.ResetBytes(&b)
+		em.Body, em.Metadata = m.Body, m.Metadata
```

Comment:
How about sending `m.Body` if `len(m.Metadata) == 0`?

That way, if I have an existing system using NATS, I can convert my senders to use Go CDK and the bytes sent over the wire will stay the same, so my existing readers can still read them.

Reply:
Will take a look, thanks.